From 54c89daddd110109a4753b113be63ba0eb6351af Mon Sep 17 00:00:00 2001
Message-Id: <54c89daddd110109a4753b113be63ba0eb6351af.1373271639.git.jdenemar@redhat.com>
From: Cole Robinson <crobinso@redhat.com>
Date: Wed, 6 Mar 2013 21:15:35 +0100
Subject: [PATCH] storage: lvm: Don't overwrite lvcreate errors

https://bugzilla.redhat.com/show_bug.cgi?id=912179

Before:
$ sudo virsh vol-create-as --pool vgvirt sparsetest --capacity 16M --allocation 0
error: Failed to create vol sparsetest
error: internal error Child process (/usr/sbin/lvchange -aln vgvirt/sparsetest) unexpected exit status 5:   One or more specified logical volume(s) not found.

After:
$ sudo virsh vol-create-as --pool vgvirt sparsetest --capacity 16M --allocation 0
error: Failed to create vol sparsetest
error: internal error Child process (/usr/sbin/lvcreate --name sparsetest -L 0K --virtualsize 16384K vgvirt) unexpected exit status 5:   Unable to create new logical volume with no extents
(cherry picked from commit 01df6f2bff98d8fc68350ab90c212780ef9db67a)
---
 src/storage/storage_backend_logical.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/storage/storage_backend_logical.c b/src/storage/storage_backend_logical.c
index b7ad88e..99b403d 100644
--- a/src/storage/storage_backend_logical.c
+++ b/src/storage/storage_backend_logical.c
@@ -699,6 +699,7 @@ virStorageBackendLogicalCreateVol(virConnectPtr conn,
 {
     int fdret, fd = -1;
     virCommandPtr cmd = NULL;
+    virErrorPtr err;
 
     if (vol->target.encryption != NULL) {
         virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
@@ -777,9 +778,11 @@ virStorageBackendLogicalCreateVol(virConnectPtr conn,
     return 0;
 
  cleanup:
+    err = virSaveLastError();
     VIR_FORCE_CLOSE(fd);
     virStorageBackendLogicalDeleteVol(conn, pool, vol, 0);
     virCommandFree(cmd);
+    virSetError(err);
     return -1;
 }
 
-- 
1.8.2.1

